heft-plugins/heft-storybook-plugin/src/StorybookPlugin.ts

```diff
@@ -0,0 +1,237 @@
+   * @remarks
+   *
+   * Storybook's conventional approach is for your app project to have direct dependencies
+   * on NPM packages such as `@storybook/react` and `@storybook/addon-essentials`.  These packages have
+   * heavyweight dependencies such as Babel, Webpack, and the associated loaders and plugins needed to
+   * build the Storybook app (which is bundled completely independently from Heft).  Naively adding these
+   * dependencies to your app's package.json muddies the waters of two radically different toolchains,
+   * and is likely to lead to dependency conflicts, for example if Heft installs Webpack 5 but
+   * `@storybook/react` installs Webpack 4.
+   *
+   * To solve this problem, `heft-storybook-plugin` introduces the concept of a separate
+   * "storykit package".  All of your Storybook NPM packages are moved to be dependencies of the
+   * storykit.  Storybook's browser API unfortunately isn't separated into dedicated NPM packages,
+   * but instead is exported by the Node.js toolchain packages such as `@storybook/react`.  For
+   * an even cleaner separation the storykit package can simply reexport such APIs.
+   */
+  storykitPackageName: string;
+
+  /**
+   * The module entry point that Heft should use to launch the Storybook toolchain.  Typically it
+   * is the path loaded the `start-storybook` shell script.
+   *
+   * @example
+   * If you are using `@storybook/react`, then the startup path would be:
+   *
+   * `"startupModulePath": "@storybook/react/bin/index.js"`
+   */
+  startupModulePath: string;
+}
+
+/** @public */
+export class StorybookPlugin implements IHeftPlugin<IStorybookPluginOptions> {
+  public readonly pluginName: string = PLUGIN_NAME;
+
+  private _logger!: ScopedLogger;
+  private _storykitPackageName!: string;
+  private _startupModulePath!: string;
+  private _resolvedStartupModulePath!: string;
+
+  /**
+   * Generate typings for Sass files before TypeScript compilation.
+   */
+  public apply(
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration,
+    options: IStorybookPluginOptions
+  ): void {
+    this._logger = heftSession.requestScopedLogger(TASK_NAME);
+
+    if (!options.storykitPackageName) {
+      throw new Error(
+        `The ${TASK_NAME} task cannot start because the "storykitPackageName"` +
+          ` plugin option was not specified`
+      );
+    }
+
+    const parseResult: IParsedPackageNameOrError = PackageName.tryParse(options.storykitPackageName);
+    if (parseResult.error) {
+      throw new Error(
+        `The ${TASK_NAME} task cannot start because the "storykitPackageName"` +
+          ` plugin option is not a valid package name: ` +
+          parseResult.error
+      );
+    }
+    this._storykitPackageName = options.storykitPackageName;
+
+    if (!options.startupModulePath) {
+      throw new Error(
+        `The ${TASK_NAME} task cannot start because the "startupModulePath"` +
+          ` plugin option was not specified`
+      );
+    }
```

Comment:
```suggestion
    if (!options.storykitPackageName) {
      this._logger.emitError(new Error((
        `The ${TASK_NAME} task cannot start because the "storykitPackageName"` +
          ` plugin option was not specified`
      );
      return;
    }

    const parseResult: IParsedPackageNameOrError = PackageName.tryParse(options.storykitPackageName);
    if (parseResult.error) {
      this._logger.emitError(new Error((
        `The ${TASK_NAME} task cannot start because the "storykitPackageName"` +
          ` plugin option is not a valid package name: ` +
          parseResult.error
      );
      return;
    }
    this._storykitPackageName = options.storykitPackageName;

    if (!options.startupModulePath) {
      this._logger.emitError(new Error((
        `The ${TASK_NAME} task cannot start because the "startupModulePath"` +
          ` plugin option was not specified`
      );
      return;
    }
```

Reply:
(See other note)